Add VOF Van Keken benchmark section

class4kayaker opened the pull request:
Adds a section to the manual where the VOF advection algorithm is used to compute the Van Keken Benchmark.

### Before your first pull request:

* [x] I have read the guidelines in our [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md) document.

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).

egpuckett commented:
@bangerth Do the tests fail because we have changed manual.tex or because ASPECT can't be built?  I believe I can build it in my local repo. @class4kayaker have you looked into this recently?

egpuckett commented:
@tjhei Could you please explain why the tests broke.  Jon may have been using a not-up-to-date master.  If I can compile ASPECT with our new Cookbook manual entry, will this fix the reasons the tests failed?

class4kayaker commented:
Fixed conflicts and rebased onto the most recent master.

@egpuckett If you click the little link next to the tests that says details, it will take you to a page with a more detailed error report. In this case, the step the tests are failing at is "Check Permissions", which checks to see if someone with the appropriate authority has authorized the pull request for testing.

egpuckett commented:
@bangerth I can understand if you are as overwhelmed with trying to put together an online class as we are.  But the PR is here if you have time. :-)

egpuckett commented:
@class4kayaker Thanks for the clarification!

naliboff commented on doc/manual/manual.tex:
```diff
@@ -8485,23 +8488,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+
+\begin{figure}[b!]
+   \centering
+\includegraphics[width=0.5\textwidth]{cookbooks/benchmarks/vof-van-keken/rms_vel_comparison.png}
```
To my eye, I can really only see the lines for refinements 5-7. Perhaps this is all that is needed to show convergence?

class4kayaker replied:
I would appreciate it @egpuckett would comment. It is true that after 7, the lines are overlapping to visual examination on the plot in question, but not including the further lines would make including that point more questionable.

egpuckett replied:
Please keep all lines / levels of refinement in.  Unless I missed something we are duplicating exactly what was done in 

> 5.4.3 The van Keken thermochemical composition benchmark

> This section is a co-production of Cedric Thieulot, Juliane Dannberg, Timo Heister and Wolfgang Bangerth with an extension to this benchmark provided by the Virginia Tech Department of Geosciences class “Geodynamics and ASPECT" co-taught by Scott King and D. Sarah Stamps.

We want to compare Apples to Apples.  We think VOF is a much better methodology **for certain problems**, but there seem to be a lot of, or maybe only some influential, ASPECT users and computational geodynamicists that think otherwise and would just as soon this methodology - interface tracking - went away.  Personally, I think that is so crazy that I can't find words to express how backward thinking this is.  Interface tracking is the right way to do certain problems in every field of mechanics, fluid mechanics, not to mention a gazillion other examples.  We need the support of the ASPECT community to implement a 3D VOF method and one that works on any kind of arbitrary grid that is supported by deal.ii.

I think the way to do that is to show everyone how much better the VOF results are **on some problems**, and this one in particular.

naliboff commented on doc/manual/manual.tex:
```diff
@@ -8485,23 +8488,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+   \caption{Evolution of the root mean square velocity as a function of time for computations 
+    of the Van Keken problem made with the VOF interface tracking algorithm with five 
+    different global mesh refinements.
+    Since the VOF initial conditions are discontinuous, the above results should really be 
```
Perhaps no need to start this sentence on a new line?

class4kayaker replied:
I had been adding linebreaks at sentence ends in my new additions relatively frequently due to the fact that latex treats that type of linebreak as standard whitespace with the intent of making it easier on me to look at the diffs.

naliboff replied:
Gotcha. Before merging, please reorganize the text into standard paragraphs and line paragraphs.

class4kayaker replied:
My understanding and testing indicates that LaTeX does not insert a paragraph break unless there is a blank line, os the current approach does not actually produce malformed output (as I indicated it combines the sentences into a paragraph as usual), which was part of the reason I used the approach in question. I can do the reformatting as a large block commit as the final change before merging though.

naliboff commented on doc/manual/manual.tex:
```diff
@@ -8485,23 +8488,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+    \centering
+    \includegraphics[width=0.4\textwidth]{cookbooks/benchmarks/vof-van-keken/vof_van_keken_refinement_comparison.png}
+    \caption{\it The results of two computations of the Van Keken problem made with the VOF 
+        interface tracking algorithm  overlaid upon each other at $t_{end}=2000$.
```
```suggestion
        interface tracking algorithm overlaid upon each other at $t_{end}=2000$.
```

naliboff commented on doc/manual/manual.tex:
```diff
@@ -8485,23 +8488,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+    \caption{\it The results of two computations of the Van Keken problem made with the VOF 
+        interface tracking algorithm  overlaid upon each other at $t_{end}=2000$.
+        This visualization consists of the \underline{two} zero contours of 
+        $\tilde{\psi}^7 \, (t_{end})$ and $\tilde{\psi}^8 \, (t_{end})$, where 
```
What exactly does `\tilde{\psi}^7` and `\tilde{\psi}^8` refer to? Reading below, my understanding is that `psi` is the zero contour of the reconstructed interface. Do `7` and `8` refer to the refinement level?

class4kayaker replied:
Yes, that is correct. I believe @egpuckett modified this to the current phrasing, while my first draft referred to the reconstructed interface, and I would therefore appreciate any comments on whether reverting to the earlier text would make more sense.

naliboff replied:
I would combine the two approaches and perhaps use something like the following wording:
```
This visualization consists of the \underline{two} zero contours of the reconstructed interface for models with refinement levels $\tilde{\psi}^7 \, (t_{end})$ and $\tilde{\psi}^8 \, (t_{end})$, where 
```

class4kayaker replied:
I checked and I had made some edits already. I will add my own note on the corresponding region of the relocated text for further discussion before marking this as resolved.

naliboff commented on doc/manual/manual.tex:
```diff
@@ -8485,23 +8488,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+        $\tilde{\psi}^7 \, (t_{end})$ and $\tilde{\psi}^8 \, (t_{end})$, where 
+        $\tilde{\psi}^7 \, (t_{end})$ is field where the zero contour is the reconstructed interface 
+        at the final time $t_{end}$ computed on a uniform grid with a refinement level of 7 and 
+        similarly for $\tilde{\psi}^8 \, (t_{end})$.
```
This sentence was quite hard to follow. Could you break into two shorter sentences and define `\psi` towards the beginning?

naliboff commented on doc/manual/manual.tex:
```diff
@@ -8485,23 +8488,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+Since, as noted above, when the interface is discontinuous, the van Keken problem is a version 
+of the Rayleigh-Taylor problem it is unstable to perturbations of all 
+wavelengths,\footnote{This is true whether the two fluids have the same viscosity or different 
+viscosities.} (e.g. see~\cite{SC:1961}), it is therefore, extremely sensitive to the initial 
```
```suggestion
viscosities.} (e.g. see~\cite{SC:1961}). Therefore, it is extremely sensitive to the initial 
```

naliboff commented on doc/manual/manual.tex:
```diff
@@ -8485,23 +8488,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+wavelengths,\footnote{This is true whether the two fluids have the same viscosity or different 
+viscosities.} (e.g. see~\cite{SC:1961}), it is therefore, extremely sensitive to the initial 
+conditions.
+Thus, rather than use the standard initialization procedure for the VOF method, which consists 
```
Can you break this down into sentences?

class4kayaker replied:
Altered in new location, will mark for comment in my own review before requesting your re-reading.

naliboff commented on doc/manual/manual.tex:
```diff
@@ -8485,23 +8488,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+First we create the function $\phi$, which has the following two properties: 1) it is positive 
+in the region that contains one of the fluids, which we will refer to as fluid~1, and negative 
+in the complement of this region, which we will refer to as fluid~2, and 2) at each point in 
+the domain the magnitude $| \phi |$ of $\phi$  is the distance to the boundary between the 
```
```suggestion
the domain the magnitude $| \phi |$ of $\phi$ is the distance to the boundary between the 
```

naliboff commented on doc/manual/manual.tex:
```diff
@@ -8485,23 +8488,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+This initialization procedure provides a much finer and thus, more accurate, initial condition 
+than the standard VOF initialization procedure described above.
+
+% AMR section, currently unused so not included
```
Is the plan to remove the following lines from the cookbook?

gassmoeller replied:
That means you have a parameter file with AMR and VoF that crashes? Please open an issue about that and post the input file there so that others who encounter this crash will see that it is a known issue.

egpuckett replied:
@class4kayaker Is it possible to run this now and see if it works? Given that deal.ii is now v. 9.2.0, perhaps it will work?

naliboff commented on doc/manual/manual.tex:
```diff
@@ -8485,23 +8488,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+\noindent
+
+We made a number of computations of the van Keken problem with the VOF method in order to 
+compare the wall clock times of the VOF results directly with the those computed with a DG
```
```suggestion
compare the wall clock times with computations using a DG compositional field.
```

naliboff commented on doc/manual/manual.tex:
```diff
@@ -8485,23 +8488,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+
+We made a number of computations of the van Keken problem with the VOF method in order to 
+compare the wall clock times of the VOF results directly with the those computed with a DG
+compositional field.
```
Delete if the suggestion above is used.

class4kayaker replied:
```suggestion

```

naliboff commented on doc/manual/manual.tex:
```diff
@@ -8485,23 +8488,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+          10 & 16           & \phantom{1}5.17 days\phantom{minutess}&>6.00 days\phantom{tess} \\
+            \hline
+        \end{tabular}
+        \caption{\it Comparison of runtimes for the van Keken problem with VOF and and a DG computational field, in which the initial conditions for DG smoothed as described in section~\ref{paragraph:van-keken compositional fields} above.
```
```suggestion
        \caption{\it Comparison of runtimes for the van Keken problem with VOF and a DG computational field, in which the initial conditions for DG smoothed are as described in section~\ref{paragraph:van-keken compositional fields} above.
```

egpuckett replied:
@class4kayaker @naliboff   I think that the second sentence here:

> ... problem with VOF and and a DG computational field, ...

the word 'computational' is meant to be 'compositional', is it not?

naliboff commented on doc/manual/manual.tex:
```diff
@@ -8485,23 +8488,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+        \label{tab:vof-runtime-comparison-table}
+\end{table}
+
+The consistency of the results shown here differs noticeably from the behavior of the  
```
In this paragraph and some above you are starting multiple sentences on new lines. Is this intentional for the purposes of readability?

class4kayaker replied:
Yes, it is intentional, but the primary purpose was to attempt to make diffs cleaner, rather than the primary intent being tied to the readability of the latex file itself.

naliboff commented on doc/manual/manual.tex:
```diff
@@ -8485,23 +8488,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+algorithms.
+One of the possible reasons for these differences is the procedure for 
+defining the initial interface.
+As described above, we have designed an approach for setting the initial condition that should 
```
Could you condense the following sentence a bit or break it into two sentences?

naliboff commented on doc/manual/manual.tex:
```diff
@@ -8485,23 +8488,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+reproduce a smooth curve to a high degree of accuracy and, in particular, the curve, which is
+where the discontinuity occurs in the discontinuous initial condition version of this problem.
+
+In order to study this feature of our algorithm, we vary the size of the initial interface 
```
```suggestion
To study this feature of our algorithm, we vary the size of the initial interface 
```

naliboff commented on doc/manual/manual.tex:
```diff
@@ -8485,23 +8488,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+where the discontinuity occurs in the discontinuous initial condition version of this problem.
+
+In order to study this feature of our algorithm, we vary the size of the initial interface 
+perturbation in order to examine the sensitivity of the final results to a small change in the 
```
```suggestion
perturbation and examine the sensitivity of the final results to a small change in the 
```

naliboff commented on doc/manual/manual.tex:
```diff
@@ -8485,23 +8488,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+
+In these computations we vary the value of $a$ from its usual value of $a = 0.02$ to 
+$5\% = 0.001$ below its usual value to $5\%$ above its usual value in increments of $0.01$.
+In other words, we compare the values for $a  =0.019$, $0.020$, and $0.021$.
```
```suggestion
In other words, we compare the values for $a =0.019$, $0.020$, and $0.021$.
```

naliboff commented on doc/manual/manual.tex:
```diff
@@ -8485,23 +8488,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+of the change in the value of $a$; i.e., $h \, \lessapprox \, \Delta a = 0.001$.
+We also note that the smoothing parameter, which would produce a
+$10^{-3}\leq C \leq 1 - 10^{-3}$ band on the order of the same size as the amplitude 
+variation shown here would be approximately $2.8 \cdot 10^{-4}$.
```
```suggestion
variation shown here, would be approximately $2.8 \cdot 10^{-4}$.
```

class4kayaker commented on cookbooks/van-keken-vof.prm:
```diff
@@ -0,0 +1,118 @@
+# Note: The compositional field is what drives the flow
+# in this example
+
+subsection Compositional fields
```
There are also some modifications to the CFL number, specifying the use of a discontinuous composition discretization, and the postprocessing step.

I did not model it on that approach due in part to the fact that the smooth vs discontinuous prm file did not use that approach either.

naliboff replied:
@class4kayaker - Given that the parameter still share many common features, I would still derive the current parameter VOF from one of the two original files. If updates are made to the benchmark later on, this ensures the current file also gets updated as well.

class4kayaker replied:
Doing a grep on all of the cookbook parameter files, I do not see any other cookbook that makes use of the approach you suggest. Additionally, it would create a dependency on the referenced file remaining in the same relative location if the cookbook is copied as a reference. As such, I am not sure that including the file is the appropriate way to set up the cookbook.

class4kayaker commented on doc/manual/manual.tex:
```diff
@@ -8204,7 +8204,7 @@ \subsubsection{Crustal deformation}
    \centering
    \includegraphics[width=0.7\textwidth]{cookbooks/crustal_deformation/grids.png}
-   \caption{\it Mesh evolution during the first timestep (refinement is based on strain rate).}
+   \caption{\it Mesh evolution during the first time step (refinement is based on strain rate).}
```
I think it was likely an extra fixup that got caught up in the rest (in either direction). I will likely remove it so the diff is cleaner

class4kayaker commented on doc/manual/manual.tex:
```diff
@@ -9033,23 +9036,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+Another advantage the VOF method has over modeling moving interface problems with compositional 
+fields is that for problems in which the interface occupies a relatively small part of the 
+computational domain, all of the computational work in the VOF method is done in cells that lie 
+in a neighborhood of the interface, rather than in the entire computational domain. 
```
Given there is vastly less work on most of the domain, the implication is not inaccurate, though it is an inherent component of the VOF algorithm, so any implementation that did not deliberately avoid using less computationally intense paths in common cases would also benefit.

egpuckett replied:
In other words ... "Yes"

Look at the table Jon made where he compared using a DG compositional field versus VOF on the same problem parameters.

class4kayaker commented on doc/manual/manual.tex:
```diff
@@ -9033,23 +9036,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+representation of the interface~\cite{JMR:2019,JMR-EGP:2019}, instead we compute the initial volume fractions using a signed distance function $\phi$ as follows.
+
+First we create the function $\phi$, which has the following two properties: 1) it is positive 
+in the region that contains one of the fluids, which we will refer to as fluid~1, and negative 
```
The tilde here is a latex based instruction to insert a space where a line break is not permitted, not a tilde directly.

class4kayaker commented on doc/manual/manual.tex:
```diff
@@ -7655,7 +7655,7 @@ \subsubsection{Crustal deformation}
 \begin{figure}
    \centering
-   \includegraphics[width=0.7\textwidth]{cookbooks/crustal_deformation/doc/grids.png}
+   \includegraphics[width=0.7\textwidth]{cookbooks/crustal_deformation/grids.png}
```
I think some other commits may have been mixed up during the rebase. I will look into correcting that.

class4kayaker commented on doc/manual/manual.tex:
```diff
@@ -7643,7 +7643,7 @@ \subsubsection{Crustal deformation}
 {\tt   set Initial adaptive refinement        = 4}
 yields a series of meshes as shown in Fig. (\ref{fig:meshes}), all produced during the 
-first timestep. As expected, we see that the location of the highest mesh refinement
+first time step. As expected, we see that the location of the highest mesh refinement
```
```suggestion
first timestep. As expected, we see that the location of the highest mesh refinement
```

class4kayaker commented on doc/manual/manual.tex:
```diff
@@ -8485,23 +8488,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
 
 This benchmark is carried out in \cite{Deu08,Ger10,thie11} and is 
-based on the analytical solution by Ramberg\cite{ramb68}, 
+based on the analytical solution by Ramberg \cite{ramb68}, 
```
Remove external modification

```suggestion
based on the analytical solution by Ramberg\cite{ramb68}, 
```

class4kayaker commented on doc/manual/manual.tex:
```diff
@@ -8485,23 +8488,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
 fluid 2 $(\rho_2,\eta_2)$ of thickness $h_2$ (with $h_1+h_2=L_y$).
 An initial sinusoidal disturbance of the interface between these
-layers is introduced and is characterised by an amplitude $\Delta$ and a
+layers is introduced and is characterized by an amplitude $\Delta$ and a
```
Remove change outside of new section

```suggestion
layers is introduced and is characterised by an amplitude $\Delta$ and a
```

class4kayaker commented on doc/manual/manual.tex:
```diff
@@ -8485,23 +8488,249 @@ \subsubsection{The van Keken thermochemical composition benchmark}
 wavelength $\lambda=L_x/2$ as shown in Figure~\ref{fig:RTi_setup}.
 
-\begin{figure}
+\begin{figure}[h!]
```
Left due to possible relevance to correct figure behavior.

gassmoeller replied:
What do you mean by `possible relevance`? Did you have to add this to create a correct figure, or can it be removed without changes? I would prefer to only add changes that are necessary to resolve issues.

bangerth replied:
Let's not use `[h!]`. It completely messes up latex's placement algorithm. The default is exactly what we should be using instead. In other words, please revert this change.

class4kayaker commented on doc/manual/manual.tex:
```diff
@@ -8485,6 +8488,231 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+defining the initial interface.
+As described above, we have designed an approach for setting the initial condition that should 
+reproduce a smooth curve to a high degree of accuracy and, in particular, the curve, which is
+where the discontinuity occurs in the discontinuous initial condition version of this problem.
```
@naliboff, @egpuckett  Would this edit work, or do you have any suggestions on how else to clarify?

```suggestion
As described above, we have designed an approach for setting the initial condition that should
reproduce the smooth curve for the initial boundary between the two fluids for this problem
to a high degree of accuracy.
```

class4kayaker commented on doc/manual/manual.tex:
```diff
@@ -8485,6 +8488,231 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+Thus, rather than use the standard initialization procedure for the VOF method, which consists 
+of integrating the initial interface over each multi-material cell in which one material or 
+fluid has the value of $1$ and the other has the value of $0$; i.e., a characteristic function 
+representation of the interface~\cite{JMR:2019,JMR-EGP:2019}, instead we compute the initial volume fractions using a signed distance function $\phi$ as follows.
```
Does this suggested edit preserve the necessary information, and make the sentence structure clearer? @egpuckett @naliboff 

```suggestion
In order to address this sensitivity, we do not use a standard composition quadrature initialization procedure
for the VOF method.
Instead we compute the initial volume fractions using a signed distance function $\phi$ as follows~\cite{JMR:2019,JMR-EGP:2019}.
```

naliboff replied:
@class4kayaker - Yes, that rephrasing is much clearer, thanks!

class4kayaker commented on doc/manual/manual.tex:
```diff
@@ -8485,6 +8488,231 @@ \subsubsection{The van Keken thermochemical composition benchmark}
+% The new mesh refinement parameter section is shown below.
+% 
+% \lstinputlisting[language=prmfile]{cookbooks/benchmarks/vof-van-keken/amr.part.prm.out}
+
```
Removal included for easier edit.

```suggestion
```

egpuckett commented:
@naliboff and @gassmoeller I have a request.  Before you merge this, please let me read the final edit.  I worked hard to get the text to sound just the way I wanted it,and I would like to have a chance to see it and comment before it is merged.  Thanks!

class4kayaker commented on cookbooks/van-keken-vof.prm:
```diff
@@ -0,0 +1,118 @@
+# A description of the van Keken et al. thermochemical composition benchmark. See the manual (section 5.43) for more
```
```suggestion
# A description of the van Keken et al. thermochemical composition benchmark. See the manual for more
```
marking change for easier batch fix

class4kayaker commented:
Changed to new structure as requested, but that lost some of the references in the change requests.

naliboff commented:
@class4kayaker - Is the text ready for another round of review?

class4kayaker commented:
I believe so, I have fixed most of the minor issues mentioned and made some attempts at including clarifications where necessary. The adjustment to the new cookbook layout does appear to have lost some of the associated notes I had been asking for suggestions on wording, so I have inserted the edits I had been thinking of for review.

naliboff commented:
/rebuild

class4kayaker commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -36,14 +36,16 @@
-        The color of the reconstructed interface zero contour $\tilde{\psi}^7
-        \, (t_{end})$ is black and the color of the zero contour
-        $\tilde{\psi}^8 \, (t_{end})$ is bright green.
+        This visualization shows the reconstructed boundary between the two
+        materials  at the final time $t_{end}$ as computed on a uniform grid
+        with 7 and 8 levels of refinement.
+        The boundaries between the materials are displayed as contours
+        of the fields $\tilde{\psi}^7\,(t_{end})$ (black) and
+        $\tilde{\psi}^8\,(t_{end})$ (bright green), which are generated by the  visualization postprocessor.
+        The contours for the reconstructed material boundaries are superimposed
+        on a color gradient visualization of the material composition for the
+        computation with 8 levels of refinement in order to make the regions
+        with each fluid type more evident.
```
Clarifying edit to description of material boundary visualization

class4kayaker commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -52,11 +54,12 @@
+In order to address this sensitivity, we do not use  the default approach of
+computing the initial material volume fractions using a composition quadrature.
+Instead we compute the initial volume fractions using a signed distance
+function $\phi$ as follows~\cite{JMR:2019,JMR-EGP:2019}.
```
Clarifying edit, requesting approval from @naliboff

class4kayaker commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -163,15 +166,17 @@
+To study this feature of our algorithm and the sensitivity of the problem to
+the precise initial condition, we vary the size of the initial interface
+perturbation and examine the sensitivity of the final results to a small change
+in the initial conditions.
```
Rewording/condensing, requesting comment from @naliboff.

naliboff commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -0,0 +1,212 @@
+problem we selected an initialization procedure specialized for the volume of
+fluid method which permits a much more consistent initialization by reducing the
+variation in the initial condition when the intial mesh is refined.
+
```
Please rephrase the sentence above to the following:

```
One possible reason for these differences is the specialized initialization procedure used for 
the volume of fluid method, which permits a much more consistent initialization by 
reducing the variation in the initial condition when the initial mesh is refined.
```

egpuckett commented:
Hi @naliboff and @class4kayaker Please don't merge until I have had a chance to read the new version.  I put a lot of time into the original one and would like to have the opportunity to suggest edits to the new version.  It will be easier for me to do this on the weekend than now.  I am virtually overwhelmed teaching a class of 95 students online, 10 or more of whom are in Asia, not to mention how do I prevent them from cheating on the first Midterm on May 1.

egpuckett commented:
P.S. @naliboff Please respond that you will hold up until I have a chance to read over the new edits.

I hope that you and yours are Safe and Healthy!

Thanks!

- Gerry

naliboff commented:
@egpuckett - sure thing.

class4kayaker commented:
I will also hold the final version on EGP's review.

class4kayaker commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -0,0 +1,212 @@
+The consistency of the results shown here differs noticeably from the behavior of the  
+problem with discontinuous initial conditions when computed with the FEM and DG advection 
+algorithms.
+One of the possible reasons for the variation in behavior between the two
+approaches is the differing procedures for defining the initial interface.
+In particular we note that for the volume of fluid version of this benchmark
+problem we selected an initialization procedure specialized for the volume of
+fluid method which permits a much more consistent initialization by reducing the
+variation in the initial condition when the intial mesh is refined.
```
```suggestion
One possible reason for these differences is the specialized initialization procedure used for 
the volume of fluid method, which permits a much more consistent initialization by 
reducing the variation in the initial condition when the initial mesh is refined.
```

class4kayaker commented:
Suggested edits made, and requesting review by @egpuckett

naliboff commented:
@egpuckett @class4kayaker - Is this ready for a final review or merge?

class4kayaker commented:
I have been waiting on @egpuckett to do a review of the text as he requested. I just checked with him again today, and he indicated that he would look it over, but I have not yet gotten any feedback from him. Assuming that he approves, I will squash down the edit onto the current master, and let you either merge or review as you think is best.

naliboff commented:
@class4kayaker - I heard from @egpuckett and he is going to go through the changes in the next few days. Thanks again for all the work on this!

egpuckett commented:
@class4kayaker There is a flaw in reference [75]

> ... in a density stratified fluid in the earthâĂŹs ...

naliboff commented:
@class4kayaker - Reading the comments from @egpuckett, it sounds like the only things left to do are (1) re-run the model that crashes with deal.II 9.2.0 and if it still crashes (2) open an issue. Is this correct?

class4kayaker commented:
I believe so, is there a reason this PR was closed?

naliboff commented:
> I believe so, is there a reason this PR was closed?

@class4kayaker - I hit "Close and comment" instead of "comment", sorry about that. Ready to merge on my end once you and @egpuckett have decided what to do with the model that crashes (open issue, retest with deal.II 9.2.0, etc).

class4kayaker commented:
I have made a similar mistake, I just wanted to check since the PR was closed.

egpuckett commented:
@naliboff @class4kayaker No, I'm sorry I'm not quite done.  As I mentioned to Jon in an email yesterday, I'm going through it conversation by conversation.  I'm sorry for the additional delay.  My son has been visiting from LA for the past 2 weeks and perhaps another week.  It has been very important for me to spend time with him, since hes is the only member of our family that has traveled since the pandemic began. That, along with the holiday, has slowed me down.  But I'll be done in a day or so.  Thanks for your patience.  

Also, @naliboff, thanks for the thorough review.  You were right. All I've found that I have suggested we change are typos.

egpuckett commented:
> > I believe so, is there a reason this PR was closed?
> 
> Ready to merge on my end once you and @egpuckett have decided what to do with the model that crashes (open issue, retest with deal.II 9.2.0, etc).

@class4kayaker @naliboff @gassmoeller My view is that it would be best if we could fix this issue now.  It doesn't look like Jon will have a PostDoc that allows him to work on ASPECT, so it would be best to clean as much up a possible now.

That said, I don't think we have to tie the issue to this PR. Do you two agree?  It doesn't look like an issue has been opened  yet.

I also would like to make sure that PR #2564 is merged an then add a cookbook.  I think the cookbook can be straight out of Jon's thesis or our Physics of Fluids paper.  The convergence test based on Richardson's Extrapolation is an extraordinarily valuable tool.

class4kayaker commented:
I have just completed a test run of the PRM file that had been erroring out on Peloton2 using a fresh compilation of both Deal.II and Aspect, and the run completed successfully. I have not yet done a comparison of the model state yet, but I believe that it will likely be very similar.

egpuckett commented:
@class4kayaker That's super!  @naliboff  & @class4kayaker is it possible to make this computation on Stampede or one of the TACC computers, since other ASPECT users in the US are more likely to have access to these machines.  @naliboff we may need your help here, but I think it will be well worth the effort!

@gassmoeller FYI regarding your comment from March 31 concerning the AMR 'crashing' issue.

class4kayaker commented:
I will look into redoing the runs on Stampede.

As far as a visual comparison, the attached image compares the AMR and uniform VK problem at the same level of refinement (7). 

![vof_van_keken_amr_comparison](https://user-images.githubusercontent.com/1050493/86625033-a64d6d80-bf81-11ea-803f-c9cb094cae53.png)

egpuckett commented:
@class4kayaker You originally planned to put this in the cookbook.  Could you still do that?  @naliboff Would that be OK.

@gassmoeller FYI

naliboff commented:
@egpuckett - yes, certainly fine to put new images or parameter files in the cookbook. However, for simplicity perhaps easier to merge this PR and make any additional significant additions with a new PR?

class4kayaker commented:
Making a separate PR would make sense, especially since I had deleted the initial text I had.

gassmoeller commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -0,0 +1,216 @@
+
+Since, as noted above, when the interface is discontinuous, the van Keken problem is a version 
+of the Rayleigh-Taylor problem that is unstable to perturbations of all 
+wavelengths,\footnote{This is true whether the two fluids have the same viscosity or different 
+viscosities.} (e.g. see~\cite{SC:1961}).
+Therefore, it is extremely sensitive to the initial conditions.
```
```suggestion
wavelengths\footnote{This is true whether the two fluids have the same viscosity or different 
viscosities.} (e.g. see~\cite{SC:1961}),
it is extremely sensitive to the initial conditions.
```

egpuckett replied:
@gassmoeller @naliboff Making a separate PR for the AMR computation and figure is fine with me.  

However, I have another question.  I have spotted one or more minor errors in **manual.bib**.  For example, in this reference

> [72] J. Ritsema and H. J. van Heijst. Seismic imaging of structural heterogeneity in earth’s mantle: ...

Earth is not capitalized here, since there needs to be curly braces around the E: {E}arth in the **manual.bib** file.

Shall I suggest the change here?  One thing to consider is that once I am done with this PR I am not likely to look at the manual or **manual.bib** for quite a while.  I have a lot on my plate at the moment.

egpuckett replied:
@gassmoeller Do you feel strongly about removing the complete sentence

> Therefore, it is extremely sensitive to the initial conditions.

and just adding

> ... it is extremely sensitive to the initial conditions.

to the end  of the previous sentence?  In other words, removing the word "Therefore, ..." as the beginning of the (complete) sentence quoted above.

I wrote those two sentences and I would prefer to remove the comma after **wavelengths** (it was definitely a mistake) but keep the period '.' after the right parentheses like this:

> (e.g. see~\cite{SC:1961}).

I think this is just a matter of writing style. :-)

class4kayaker replied:
I have modified the text as suggested in @egpuckett's comment, and would appreciate a response on whether I should look into further modifications. I will note that to me the phrasing suggested by EGP is more natural, as it includes the transitioning structure. Some rephrasing might be appropriate, but just removing the word "Therefore" doesn't feel correct.

gassmoeller replied:
I am perfectly fine with your solution, I just wanted to fix the additional comma in some form.

class4kayaker commented on cookbooks/.gitignore:
```diff
@@ -0,0 +1 @@
+**/doc/*.out
```
Yes that does handle this, I have deleted this fille.

egpuckett commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -0,0 +1,216 @@
+\end{figure}
+
+Since, as noted above, when the interface is discontinuous, the van Keken problem is a version 
+of the Rayleigh-Taylor problem that is unstable to perturbations of all 
```
```suggestion
of the Rayleigh-Taylor problem, which is unstable to perturbations of all 
```
I believe this is a case where the rules of grammar specify that one should use the word 'which' rather than 'that'.

The comma preceding the word 'which' is part of Cambridge University Press' "Guide for Authors".

class4kayaker replied:
Fixed and commited.

class4kayaker commented:
If there are no additional changes to be made from EGP's review, I have squashed the commit chain, and it is ready to merge upon approval.

egpuckett commented:
I do want to look at a few more things.  My apologies for the delay.

egpuckett commented:
@naliboff Please merge this PR, if you are still happy with it. Thanks!

bangerth commented on cookbooks/van-keken-vof/doc/amr.part.prm:
```diff
@@ -0,0 +1,8 @@
+subsection Mesh refinement
+  set Initial adaptive refinement        = 0
+  set Strategy                           = volume of fluid interface
+  set Initial global refinement          = 7
```
That's quite a fine mesh. How long does the cookbook run? We try to keep the cookbooks so that they run in a reasonable amount of time by default, and if the mesh is too coarse to be accurate, then comment on that in the cookbook and possibly show results from a more refined mesh.

egpuckett replied:
@bangerth I believe we always ran the same meshes that Scott King, et al. ran for compositional fields, never finer.  VOF should always be less expensive than compositional fields on a problem with a well defined interface like this one.

class4kayaker replied:
With 4 processors on Peloton2, the run took 1.15 hrs. I could adjust the version in the cookbook to 6 levels of refinement which took ~8 minutes if a 1hr runtime is too long.

bangerth replied:
Yes, let's do that. We try to set the defaults in the cookbooks in such a way that things can be run on a laptop with reasonable effort so that people can immediately start to play.

bangerth commented on cookbooks/van-keken-vof/doc/main.part.prm:
```diff
@@ -0,0 +1,20 @@
+  subsection Function
+    set Variable names      = x,z
+    set Function constants  = pi=3.1415926
+    set Function expression  =  0.2+0.02*cos(pi*x/0.9142)-z
```
might be nice to align the `=` signs :-)

bangerth commented on cookbooks/van-keken-vof/doc/variation.part.prm:
```diff
@@ -0,0 +1,10 @@
+  subsection Function
+    set Variable names      = x,z
+    set Function constants  = pi=3.1415926, a=0.02
+    set Function expression  =  0.2+a*cos(pi*x/0.9142)-z
```
same here with aligning `=`

bangerth commented on doc/manual/manual.tex:
```diff
@@ -8286,6 +8289,12 @@ \subsubsection{The van Keken thermochemical composition benchmark}
 We can observe that when the two-layer viscosity model is added, there is only one apparent peak for each case. The first peaks of the 0.01 viscosity contrast tests appear earlier and are larger in magnitude than those of 0.1 viscosity contrast tests.  There are no secondary plumes and the whole system tends to reach stability after around 500 time steps.
 
+
+\paragraph{Computation of the van Keken Problem with the Volume-of-Fluid Interface Tracking Method}
```
I think that you should just make this a `\subsubsection` like the other benchmarks/cookbooks. You should also put the note at the top that documents who contributed the section (i.e., *you*).

egpuckett replied:
@bangerth This was a collaboration between @class4kayaker and me.  He wrote the first draft and made the computations.  I wrote the final draft.

bangerth replied:
@egpuckett -- just add something like that as the first sentence of the subsubsection, like is done for most of the other subsubsections.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -8058,6 +8058,8 @@ \subsubsection{The van Keken thermochemical composition benchmark}
 Timo Heister and Wolfgang Bangerth with an extension to this benchmark provided by the Virginia Tech Department of Geosciences class ``Geodynamics and ASPECT'' co-taught by Scott King and D.~Sarah Stamps.}
 
+\paragraph{Computation of the van Keken Problem with compositional fields}
+\label{paragraph:van-keken compositional fields}
```
Just drop the `\paragraph` and use the label from the subsubsection instead (see also my comment below about the new paragraph you add).

egpuckett commented:
@bangerth If you scroll up a bit you'll notice the output from a computation of the van Keken benchmark with AMR.  It is not currently part of this PR.  I think @gassmoeller suggested we merge this PR and then make a separate PR for the AMR computation. Are you OK with that?   The reason we put an AMR computation in the cookbook is that VOF is especially well suited to having the finest grid **only** over the interface that is being tracked by the VOF algorithm.  I think that it is important to emphasize this to future users and perhaps to the community in general (so they might be tempted to try it.) This cuts down on the time of the computation considerably, while simultaneously keeping the interface resolved as accurately as the user wants it to be.

bangerth commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -0,0 +1,216 @@
+In particular, the combination of \texttt{Number of initialization samples} with the 
+\texttt{level set} initialization type indicates that our initialization will consist of 
+dividing each grid cell into $16 \times 16$ subcells and the distance to the given initial 
+interface $f(x)$, provided in \texttt{Function expression}, is computed in each of the 256 
```
```suggestion
interface $f(\mathbf x)$, provided in \texttt{Function expression}, is computed in each of the 256 
```

bangerth commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -0,0 +1,216 @@
+dividing each grid cell into $16 \times 16$ subcells and the distance to the given initial 
+interface $f(x)$, provided in \texttt{Function expression}, is computed in each of the 256 
+subcells.
+We then use this information to compute a piecewise linear interface approximation to $f(x)$.
```
```suggestion
We then use this information to compute a piecewise linear interface approximation to $f(\mathbf x)$.
```

bangerth commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -0,0 +1,216 @@
+One can also model the van Keken problem with the volume-of-fluid (VOF) interface tracking 
+algorithm in ASPECT.
```
Most readers will probably not know about the VoF method. It might be useful to add a few references here where one could look up what that term means.

bangerth commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -0,0 +1,216 @@
+with fluid~1, and negative in the region with fluid~2.
+However $\tilde{\psi}$ does not satisfy the requirement that the magnitude is
+equal to the distance to the interface as would be required for the signed
+distance function $\tilde{\phi}$.
```
did you mean
```suggestion
distance function $\phi$.
```
?

bangerth commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -0,0 +1,216 @@
+The modifications to the parameter file that are necessary in order to draw the reconstructed 
+boundary as a contour are shown immediately below.
+The full configuration file for this version of the benchmark problem can be found at 
+\url{cookbooks/van-keken-vof.prm}.
```
I think you're missing a directory name under `cookbooks/` here

class4kayaker replied:
```suggestion
\url{cookbooks/van-keken-vof/van-keken-vof.prm}.
```
Comment addeded to ease use of github suggestion batch.

bangerth commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -0,0 +1,216 @@
+We made a number of computations of the van Keken problem with the VOF method in order to 
+compare the wall clock times with computations using a DG compositional field.
+We ran both on the same cluster at global refinements 5--8 using one node with four CPUs and 
+9--10 using two nodes with 16 CPUs.
```
```suggestion
refinements 9 and 10 using two nodes with 16 CPUs.
```

bangerth commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -0,0 +1,216 @@
+     \hline
+      Global Refinement & Number of Processors & VOF & DG \\
+      \hline
+\phantom{1}5 & \phantom{1}4 &\phantom{6}1.33 minutes\phantom{hours} &\phantom{1}2.57 minutes  \\
```
You won't need the `\phantom` part in the first two columns if you simply make them right-aligned in the `{|c|c|c|c|}` description above -- but you can also just leave it as is given that it shows nicely in the current version

egpuckett replied:
I'm in favor of leaving it as is.  (I wrote it. :-)

bangerth commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -0,0 +1,216 @@
+            and a DG compositional field, in which the initial conditions for
+            DG smoothed are as described in section~\ref{paragraph:van-keken
+            compositional fields} above.  The times shown are for the full
+            computation, ending at $t_{end} = 2000$ with a CFL number of
```
```suggestion
            computation, ending at $t_\text{end} = 2000$ with a CFL number of
```

bangerth commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -0,0 +1,216 @@
+            computation, ending at $t_{end} = 2000$ with a CFL number of
+            $\sigma=0.5$ in both cases. 
+            All of these computations were made with ASPECT version 2.2.0-pre
+            (master, ef542ecc2) in release mode on the Peloton2 cluster at
```
```suggestion
            (master, commit \texttt{ef542ecc2}) in release mode on the Peloton2 cluster at
```

bangerth commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -0,0 +1,216 @@
+            U.C.~Davis.
+            We note that the change in the CFL number $\sigma$ and the
+            differing choice of cluster makes a direct quantitative comparison
+            between this table and table~\ref{tab:runtime-table} invalid due to
```
```suggestion
            between this table and Table~\ref{tab:runtime-table} invalid due to
```

bangerth commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -0,0 +1,216 @@
+\begin{figure}[htb]
+    \centering
+    \includegraphics[width=0.5\textwidth]{cookbooks/van-keken-vof/doc/init_diff_rms_vel_comparison.png}
+    \caption{\it Computations of the Van Keken problem made with the VOF
```
```suggestion
    \caption{\it Computations of the van Keken problem made with the VOF
```

bangerth commented:
@egpuckett -- yes, let's get this merged first before tackling AMR. One step at a time. AMR can be addressed in a follow-up PR.

class4kayaker commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -0,0 +1,216 @@
+First we create the function $\phi$, which has the following two properties: 1) it is positive 
+in the region that contains one of the fluids, which we will refer to as fluid~1, and negative 
+in the complement of this region, which we will refer to as fluid~2, and 2) at each point in 
+the domain the magnitude $| \phi |$ of $\phi$ is the distance to the boundary between the 
```
By definition, a signed distance function (which is the term for the type of function we define here) requires that it be the actual distance. In practice, the use of the horizontal or vertical distance is often close enough when considering small perturbations, but this is a case of a sufficiently close approximation rather than them actually satisfying the requirements for a signed distance function.

bangerth replied:
Yes, I think below you make that distinction. Feel free to do with my comment whatever seems right to you.

class4kayaker replied:
I added some text below for the purpose of trying to make the use clearer.

class4kayaker commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -0,0 +1,216 @@
+Since, as noted above, when the interface is discontinuous, the van Keken problem is a version 
+of the Rayleigh-Taylor problem, which is unstable to perturbations of all 
+wavelengths\footnote{This is true whether the two fluids have the same viscosity or different 
+viscosities.} (e.g. see~\cite{SC:1961}).
```
Attempted fix made

egpuckett replied:
Looks good to me. - I wrote the original version.  @bangerth You're right, the "Since" was a bit too much. :-)

egpuckett commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -0,0 +1,216 @@
+in the complement of this region, which we will refer to as fluid~2, and 2) at each point in 
+the domain the magnitude $| \phi |$ of $\phi$ is the distance to the boundary between the 
+two fluids or materials.
+In the computations shown here, we use an approximation $\tilde{\phi}$ to $\phi$ such that   
```
```suggestion
In the computations shown here, we use an approximation $\tilde{\phi}$ to $\phi$ such that the
```

class4kayaker commented on cookbooks/van-keken-vof/doc/van-keken-vof.tex:
```diff
@@ -0,0 +1,215 @@
+two fluids or materials.
+In the computations shown here, we use an approximation $\tilde{\phi}$ to $\phi$ such that   
+difference between $\tilde{\phi}$ and $\phi$ is small enough for the purposes of making the 
+computations high-quality.
```
```suggestion
computations high-quality.
The use of an approximation as opposed to the function itself is due to obtaining an appropriate function for almost all nontrivial (not a line or a circle) boundaries is extremely difficult.
In this case, because the boundary is close to horizontal, we use the vertical distance to the boundary based on the argument that the gradient will not differ sufficiently from $1$ to induce errors in the initialization computation.
```
Attempted elaboration to clari

egpuckett replied:
@class4kayaker The gradient is a vector, not a scalar.  I think that you meant 

> ... the gradient of $\phi$ will not differ sufficiently from $(1,0)$ so as to introduce large errors in the initialization computation.

egpuckett commented:
@bangerth @naliboff @gassmoeller I believe that this is ready to be merged.  Thanks!

naliboff commented:
@bangerth - Would you like any additional changes on this PR or is this ready to go from your end?
